tests/integration/construct_component_provider/testcomponent-go/main.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2016-2021, Pulumi Corporation.  All rights reserved.
+
+package main
+
+import (
+	"fmt"
+
+	"github.com/blang/semver"
+
+	"github.com/pulumi/pulumi/pkg/v3/resource/provider"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+	pulumiprovider "github.com/pulumi/pulumi/sdk/v3/go/pulumi/provider"
+)
+
+type Provider struct {
+	pulumi.ProviderResourceState
+
+	Message pulumi.StringOutput `pulumi:"message"`
+}
+
+type Component struct {
+	pulumi.ResourceState
+
+	Message pulumi.StringOutput `pulumi:"message"`
+}
+
+func NewComponent(ctx *pulumi.Context, name string, opts ...pulumi.ResourceOption) (*Component, error) {
+	component := &Component{}
+	err := ctx.RegisterComponentResource("testcomponent:index:Component", name, component, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Test that we're indeed getting back an instance of `Provider` with its state.
+	provider := component.GetProvider("testcomponent::").(*Provider)
+	component.Message = provider.Message
```

Comment:
This sort of assignment is implied in base constructor call in other langs?

Reply:
The other langs do it explicitly in their constructors:

Python:

https://github.com/pulumi/pulumi/pull/7624/files#diff-357f441447e8aab4d038abbb5a01d00141a8e9d20d5651e8fc9dae378a075f7dR46

```python
         provider = self.get_provider("testcomponent::")
         if not isinstance(provider, Provider):
             raise Exception(f"provider is not an instance of Provider: {provider}")
         self.message = provider.message
         self.register_outputs({
             "message": self.message,
         })
```

Node.js:

https://github.com/pulumi/pulumi/pull/7624/files#diff-88176009303a404a3c6fa167d388546cd26c977e414ff1b342afe5ad0b8cf4d1R28

```typescript
         const provider = this.getProvider("testcomponent::");
         if (!(provider instanceof Provider)) {
             throw new Error("provider is not an instance of Provider");
         }
         this.message = provider.message;
         this.registerOutputs({
             message: provider.message,
         });
```

Node.js differs slightly with it happening in `initialize` rather than the constructor itself, following the pattern from #3676.